cupy/_core/dlpack.pyx

```diff
@@ -20,138 +15,213 @@ import cupy
+    cdef int64_t* shape_strides = <int64_t*>stdlib.malloc(
+        ndim * sizeof(int64_t) * 2)
+    if shape_strides == NULL:
+        stdlib.free(dlm_tensor_ptr)
+        raise MemoryError()
+
+    # We need a different setup for versioned/unversioned when it comes to
+    # the context/deleter and additional info in the newer versioned one.
+    if use_versioned:
+        dlm_tensor_ver = <DLManagedTensorVersioned*>dlm_tensor_ptr
+
+        # dl_tensor is identically filled for versioned and unversioned:
+        dl_tensor = &dlm_tensor_ver.dl_tensor
+        capsule_name = CAPSULE_NAME_VER
+
+        dlm_tensor_ver.manager_ctx = <void*>owner
+        cpython.Py_INCREF(owner)
+        dlm_tensor_ver.deleter = deleter_ver
+
+        # "Versioned" specific initialization:
+        dlm_tensor_ver.version.major = 1
+        dlm_tensor_ver.version.minor = 0
+
+        # CuPy arrays are writeable but may be copied if copying to the CPU.
+        dlm_tensor_ver.flags = 0
+        if owner is not array:
+            dlm_tensor_ver.flags |= DLPACK_FLAG_BITMASK_IS_COPIED
+    else:
+        dlm_tensor = <DLManagedTensor*>dlm_tensor_ptr
+        # dl_tensor is identically filled for versioned and unversioned:
+        dl_tensor = &dlm_tensor.dl_tensor
+        capsule_name = CAPSULE_NAME
+
+        dlm_tensor.manager_ctx = <void*>owner
+        cpython.Py_INCREF(owner)
+        dlm_tensor.deleter = deleter
+
+    # Create capsule now.  After this, the capsule will clean up on error.
+    # (Note that it is good to handle expected BufferErrors early.)
+    try:
+        capsule = cpython.PyCapsule_New(
+            dlm_tensor_ptr, capsule_name, pycapsule_deleter)
+    except BaseException:
+        stdlib.free(dlm_tensor)
+        stdlib.free(shape_strides)
+        cpython.Py_DECREF(array)
+        raise
+
+    # Fill in and finalize the dl_tensor struct (need shape set for cleanup)
+    dl_tensor.ndim = ndim
+    dl_tensor.shape = shape_strides
+    dl_tensor.strides = shape_strides + ndim
+    dl_tensor.byte_offset = 0
+
+    dl_tensor.dtype.code = dtype_code
+    dl_tensor.dtype.lanes = <uint16_t>1
+    dl_tensor.dtype.bits = <uint8_t>(dtype_itemsize * 8)
+
+    dl_tensor.device = device
+
+    # Fill in the shape and strides information (depends on GPU vs CPU array).
+    # (assumes strides are a multiple of itemsize, that should be OK for cupy.)
+    if owner is array:
+        dl_tensor.data = <void *>array.data.ptr
+
+        for n in range(ndim):
+            shape_strides[n] = array._shape[n]
+
+        for n in range(ndim):
+            shape_strides[n + ndim] = array._strides[n] // dtype_itemsize
     else:
-        raise ValueError('Unknown dtype')
-    dtype.lanes = <uint16_t>1
-    dtype.bits = <uint8_t>(array.dtype.itemsize * 8)
+        # Same as above, but we got a NumPy array, so go through Python
+        # in the off-chance that the copy has changed the strides.
+        dl_tensor.data = <void *><intptr_t>owner.ctypes.data
```

Comment:
nit: I might be paranoid but I feel I've been bitten by operator precedence before...
```suggestion
        dl_tensor.data = <void *><intptr_t>(owner.ctypes.data)
```